logins-api/android/library/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt

```diff
@@ -8,208 +8,181 @@
         }
     }
 
-    // This says it's unused but apparently this is how you add a finalizer in kotlin.
-    // No override or anything
-    fun finalize() {
-        this.close()
+    // In practice we usually need to be synchronized to call this safely, so it doesn't
+    // synchronize itself
+    private inline fun <U> nullableRustCall(callback: (RustError.ByReference) -> U?): U? {
+        val e = RustError.ByReference()
+        try {
+            val ret = callback(e)
+            if (e.isFailure()) {
+                throw e.intoException()
+            }
+            return ret
+        } finally {
+            // This only matters if `callback` throws (or does a non-local return, which
+            // we currently don't do)
+            e.ensureConsumed()
+        }
     }
 
-    companion object {
+    private inline fun <U> rustCall(callback: (RustError.ByReference) -> U?): U {
+        return nullableRustCall(callback)!!
+    }
 
-        internal fun getAndConsumeString(p: Pointer?): String? {
-            if (p == null) {
-                return null;
-            }
-            try {
-                return p.getString(0, "utf8");
-            } finally {
-                PasswordSyncAdapter.INSTANCE.sync15_passwords_destroy_string(p);
-            }
+    private inline fun <U> nullableRustCallWithLock(callback: (RawLoginSyncState, RustError.ByReference) -> U?): U? {
+        return synchronized(this) {
+            checkUnlocked()
+            val raw = this.raw.get()!!
+            nullableRustCall { callback(raw, it) }
         }
+    }
 
-        internal fun <U> safeAsync(callback: (RustError.ByReference) -> U): SyncResult<U> {
-            val result = SyncResult<U>()
-            val e = RustError.ByReference()
-            launch {
-                synchronized(PasswordSyncAdapter.INSTANCE) {
-                    val ret: U;
-                    try {
-                        ret = callback(e)
-                    } catch (e: Exception) {
-                        result.completeExceptionally(e)
-                        return@launch
-                    }
-                    if (e.isFailure()) {
-                        result.completeExceptionally(e.intoException())
-                    } else {
-                        result.complete(ret)
-                    }
-                }
-            }
-            return result
-        }
+    private inline fun <U> rustCallWithLock(callback: (RawLoginSyncState, RustError.ByReference) -> U?): U {
+        return nullableRustCallWithLock(callback)!!
+    }
+}
 
-        internal fun safeAsyncString(callback: (RustError.ByReference) -> Pointer?): SyncResult<String?> {
-            return safeAsync { e -> getAndConsumeString(callback(e)) }
-        }
+/**
+ * Helper to read a null terminated String out of the Pointer and free it.
+ *
+ * Important: Do not use this pointer after this! For anything!
+ */
+internal fun Pointer.getAndConsumeString(): String {
```

Comment:
nit: Maybe `getAndConsume*Rust*String`, just to be parallel?

Can you think of names that more closely map to Rust's ownership model?  This is really "take ownership" or "consume"; `getRustString` is really "clone" (right?)?

Reply:
I'll rename it to getAndConsumeRustString. Fxa has a similar set of functions, and we can probably think of better names when/before we figure out a way to share them.